src/server/service/kv.rs

```diff
@@ -1138,6 +753,784 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
 
         ctx.spawn(future);
     }
+
+    fn batch_commands(
+        &mut self,
+        ctx: RpcContext,
+        stream: RequestStream<BatchCommandsRequest>,
+        sink: DuplexSink<BatchCommandsResponse>,
+    ) {
+        let (tx, rx) = unbounded(GRPC_MSG_NOTIFY_SIZE);
+        let executor = self.collect_runtime.executor();
+
+        let ctx = Arc::new(ctx);
+        let peer = ctx.peer();
+        let storage = self.storage.clone();
+        let cop = self.cop.clone();
+        let thread_load = Arc::clone(&self.thread_load);
+
+        let request_handler = stream.for_each(move |mut req| {
+            let request_ids = req.take_request_ids().into_iter();
+            let requests = req.take_requests().into_vec();
+            GRPC_REQ_BATCH_COMMANDS_SIZE.observe(requests.len() as f64);
+            for (id, req) in request_ids
+                .zip(requests)
+                .filter_map(|(i, req)| req.cmd.map(|r| (i, r)))
+            {
+                handle_batch_commands_request(
+                    &storage,
+                    &cop,
+                    peer.clone(),
+                    &executor,
+                    id,
+                    req,
+                    tx.clone(),
+                    Arc::clone(&thread_load),
+                );
+            }
+            future::ok::<_, _>(())
+        });
+
+        ctx.spawn(request_handler.map_err(|e| error!("batch_commands error: {}", e)));
+
+        let thread_load = Arc::clone(&self.thread_load);
+        let response_retriever = BatchReceiver::new(
+            rx,
+            GRPC_MSG_MAX_BATCH_SIZE,
+            BatchCommandsResponse::new,
+            |batch_resp, (id, resp)| {
+                batch_resp.mut_request_ids().push(id);
+                batch_resp.mut_responses().push(resp);
+            },
+        );
+
+        let response_retriever = response_retriever
+            .inspect(|r| GRPC_RESP_BATCH_COMMANDS_SIZE.observe(r.request_ids.len() as f64))
+            .map(move |mut r| {
+                r.set_transport_layer_load(thread_load.load() as u64);
+                (r, WriteFlags::default().buffer_hint(false))
+            })
+            .map_err(|e| {
+                let code = RpcStatusCode::Unknown;
+                let msg = Some(format!("{:?}", e));
+                GrpcError::RpcFailure(RpcStatus::new(code, msg))
+            });
+
+        ctx.spawn(
+            sink.send_all(response_retriever)
+                .map(|_| ())
+                .map_err(|e| debug!("{} failed: {:?}", "coprocessor_stream", e)),
+        );
+    }
+}
+
+fn response_batch_commands_request<F>(
+    executor: &TaskExecutor,
+    id: u64,
+    resp: F,
+    tx: Sender<(u64, BatchCommandsResponse_Response)>,
```

Comment:
`BatchCommandsResponse_Response` looks wired

Reply:
I agree but it's generated by protobuf. :(